### ISSUE TITLE
Release the graph-based module ?

### ISSUE THREAD

**Issue by DebangLi** [endernewton/iter-reason#7]
@endernewton Hi, xinlei. I like your paper very much, but I have some confusion about the graph based reasoning?  Could you release the graph-based global reasoning code?

**Comment by endernewton** [comment]
yep, we will release the spatial relationships part in the next month or so. :)

**Comment by drbingjiexu** [comment]
@endernewton Hi, Xinlei, I also would like to clarify the graph-based global reasoning. Look forward to your releasing of this module code. Thanks :)

**Comment by jhneo** [comment]
@endernewton  I'm  also looking forward to the part of global reasoning with knowledge graph

**Comment by ruotianluo** [comment]
following up.

**Comment by mbuet2ner** [comment]
@endernewton just a follow up on the requests. I am really interested in your implementation as it is very relevant for my thesis. Thanks :)

See also #10 #6

**Comment by endernewton** [comment]
Sorry about the real delay on this, as I really need to come back to do this but haven't got time to (essentially go back to TF as I am mostly using C2/PT now). I am first sharing the raw code (dropbox links) from previous implementations here as they are not working on my recently installed TF, and help (if someone can) to incorporate them to the main code would be appreciated. 

https://www.dropbox.com/s/b15qdzpevbj12o4/gattend_memory.py?dl=0
https://www.dropbox.com/s/cwnd4wto1iil1ms/gbase_memory.py?dl=0
https://www.dropbox.com/s/9c71yp8g717w2be/snippets.py?dl=0

**Comment by ZhangLin-PKU** [comment]
Does anyone complete the incorporation?

**Comment by ZhangLin-PKU** [comment]
@endernewton hi, can you provide the config.py corresponding to the new graph version (which includes the GRA variable)? Thanks a lot for you help :)

**Comment by JeremyYR** [comment]
I can try it. But utils.snippets.build_region_graph, normalize and config are needed, can you share them?

**Comment by ZhangLin-PKU** [comment]
@JeremyYR The build_region_graph and normalize functions are already included in the latest snippets.py the author released (https://www.dropbox.com/s/9c71yp8g717w2be/snippets.py?dl=0). I tried to rewrite the config.py to contain the GRA variable, here is the code [config.py](https://github.com/Irene-software/important-files/blob/master/config.py). 

All attributes of GRA in gattend_memory.py and gbase_memory.py are added, in the mean time three MEM attributes CC, SAT and INPUT which are new MEM attributes are also added. The MEM.ITER is changes to 3 as declared in section 4.3. I have written all comments of these attributes which explain their meanings and functions. Please note that the GRA.W_B for vg dataset is not contained, but it corresponds to the top 10 most frequent relationships as explained in section 4.1.

I hope this can help you. If you have any new progress, please tell me. Thanks a lot!

**Comment by JeremyYR** [comment]
@Irene-software Thanks a lot. I will :)

**Comment by mjantoun** [comment]
Where did you get the ade-vec.pkl file from? When I downloaded the ADE dataset I only got index_ade20k.mat file.

**Comment by Circle-Liu** [comment]
gbase_memory.py（https://www.dropbox.com/s/cwnd4wto1iil1ms/gbase_memory.py?dl=0）include two *.csv files which are 'data/vg-graph/%s.csv' and 'data/ade-graph/%s.csv'. and where can I find them?
